### PR TITLE
verbs/RDM. Fabtests enable

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
tested with fi_rdm_tagged_pingpong:
1) modification of the test: hints->mode = FI_LOCAL_MR | FI_CONTEXT;
2) server/host argument (-s) should have valid IP address of ib device

@a-ilango @shefty

### CODE REVIEW

**Comment by sayantansur** [comment]
I was also going to add FI_CONTEXT support to the fabests benchmarks.

**Comment by evgeny-leksikov** [comment]
@sayantansur I have not upstreamed modification of the test yet. If you are preparing similar patch, go ahead.

**Comment by sayantansur** [comment]
I don't have it prepared yet :-) If you have one, that will be greatly appreciated!

**Comment by evgeny-leksikov** [comment]
Not yet :) Should it be only tagged benchmarks or something else?

**Comment by sayantansur** [comment]
The tagged/RDM ones are the ones I'm interested in. Thanks!

**Comment by evgeny-leksikov** [comment]
Related [PR to fabtests](https://github.com/ofiwg/fabtests/pull/494)

File: prov/verbs/src/fi_verbs.c [hunk 1]
```diff
@@ -80,7 +145,7 @@ int fi_ibv_create_ep(const char *node, const char *service,
 		goto out;
 
 	if (!node && !rai_hints.ai_dst_addr) {
-		if (!rai_hints.ai_src_addr && !service) {
+		if (!rai_hints.ai_src_addr) {
```
Comment (a-ilango) [inline prov/verbs/src/fi_verbs.c]: This change prevents listening on wildcard address when the app specifies just the port but no source address info.

File: prov/verbs/src/fi_verbs.c [hunk 2]
```diff
@@ -67,6 +67,71 @@ int fi_ibv_sockaddr_len(struct sockaddr *addr)
...
+	if (!cm->ec) {
+		VERBS_INFO(FI_LOG_EP_CTRL,
+			"Failed to create listener event channel: %s\n",
+			strerror(errno));
+		return -FI_EOTHER;
+	}
+
+	int fd = cm->ec->fd;
+	int flags = fcntl(fd, F_GETFL, 0);
+	int ret = fcntl(fd, F_SETFL, flags | O_NONBLOCK);
```
Comment (a-ilango) [inline prov/verbs/src/fi_verbs.c]: fi_fd_nonblock function from common code can be used here.

File: prov/verbs/src/fi_verbs.c [hunk 3]
```diff
@@ -67,6 +67,71 @@ int fi_ibv_sockaddr_len(struct sockaddr *addr)
...
+		return -FI_EOTHER;
+	}
+
+	if (rdma_create_id(cm->ec, &cm->listener, NULL, RDMA_PS_TCP)) {
+		VERBS_INFO(FI_LOG_EP_CTRL, "Failed to create cm listener: %s\n",
+			     strerror(errno));
+		return -FI_EOTHER;
+	}
+
+	if (fi_ibv_rdm_tagged_find_ipoib_addr((struct sockaddr_in*)rai->ai_src_addr,
```
Comment (a-ilango) [inline prov/verbs/src/fi_verbs.c]: Is this check needed here? rai contains the address returned from rdma_getaddrinfo. I believe that should contain the address that belongs to one of the IPoIB interfaces.

File: prov/verbs/src/ep_rdm/verbs_utils.c [hunk 4]
```diff
@@ -114,3 +115,60 @@ void fi_ibv_rdm_conn_init_cm_role(struct fi_ibv_rdm_tagged_conn *conn,
...
 }
+
+/* find the IPoIB address of the device opened in the fi_domain call. The name
+ * of this device is _domain->verbs->device->name. The logic of the function is:
+ * iterate through all the available network interfaces, find those having "ib"
+ * in the name, then try to test the IB device that correspond to each address.
+ * If the name is the desired one then we're done.
+ */
+int fi_ibv_rdm_tagged_find_ipoib_addr(const struct sockaddr_in *addr,
+				      struct ibv_context *ctx,
```
Comment (a-ilango) [inline prov/verbs/src/ep_rdm/verbs_utils.c]: ctx is not used anywhere in this function.

File: prov/verbs/src/fi_verbs.c (hunk 1 above)
Comment (evgeny-leksikov) [inline prov/verbs/src/fi_verbs.c]: Fixed

File: prov/verbs/src/fi_verbs.c (hunk 3 above)
Comment (evgeny-leksikov) [inline prov/verbs/src/fi_verbs.c]: Unfortunately that contains zeros for local host

File: prov/verbs/src/ep_rdm/verbs_rdm.h [hunk 5]
```diff
@@ -33,20 +33,22 @@
...
-        *((unsigned char*) addr), *((unsigned char*) addr + 1),	   \
-        *((unsigned char*) addr + 2),*((unsigned char*) addr + 3), \
-        *((unsigned char*) addr + 4),*((unsigned char*) addr + 5)
+#define FI_IBV_RDM_ADDR_STR(addr)				\
+        *((unsigned char*) addr + sizeof(sa_family_t) + 0),	\
+	*((unsigned char*) addr + sizeof(sa_family_t) + 1),	\
+	*((unsigned char*) addr + sizeof(sa_family_t) + 2),	\
+	*((unsigned char*) addr + sizeof(sa_family_t) + 3),	\
+	*((unsigned char*) addr + sizeof(sa_family_t) + 4),	\
+	*((unsigned char*) addr + sizeof(sa_family_t) + 5)
```
Comment (shefty) [inline prov/verbs/src/ep_rdm/verbs_rdm.h]: This is kind of ugly.  I think we need a common call for converting a sockaddr into a string -- one that uses inet_ntoa.  There may be one in the code already somewhere.

**Comment by a-ilango** [comment]
:+1:

**Pull request merged**